internal/dag/gatewayapi_processor.go

```diff
@@ -195,39 +196,121 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
 	return strings.ToLower(certificateRef.Kind) == "secret" && strings.ToLower(certificateRef.Group) == "core"
 }
 
-func (p *GatewayAPIProcessor) computeHosts(route *gatewayapi_v1alpha1.HTTPRoute) ([]string, []error) {
-	// Determine the hosts on the route, if no hosts
-	// are defined, then set to "*".
+// computeHosts validates the hostnames for a HTTPRoute as well as validating
+// that the hostname on the HTTPRoute matches what is optionally defined on the
+// listener.hostname.
+func (p *GatewayAPIProcessor) computeHosts(hostnames []gatewayapi_v1alpha1.Hostname, listenerHostname *gatewayapi_v1alpha1.Hostname) ([]string, []error) {
+
 	var hosts []string
 	var errors []error
-	if len(route.Spec.Hostnames) == 0 {
+
+	// Determine the hosts on the hostnames, if no hosts
+	// are defined, then set to "*". If the listenerHostname is defined,
+	// then the hostnames must match the Gateway hostname.
+	if len(hostnames) == 0 && listenerHostname == nil {
 		hosts = append(hosts, "*")
 		return hosts, nil
 	}
 
-	for _, host := range route.Spec.Hostnames {
+	if listenerHostname != nil {
+		if string(*listenerHostname) != "*" {
+
+			// Validate listener hostname.
+			if err := validHostName(string(*listenerHostname)); err != nil {
+				return hosts, []error{err}
+			}
+
+			if len(hostnames) == 0 {
+				hosts = append(hosts, string(*listenerHostname))
+				return hosts, nil
+			}
+		}
+	}
+
+	for _, host := range hostnames {
 
 		hostname := string(host)
-		if isIP := net.ParseIP(hostname) != nil; isIP {
-			errors = append(errors, fmt.Errorf("hostname %q must be a DNS name, not an IP address", hostname))
+
+		// A "*" hostname matches anything.
+		if hostname == "*" {
+			hosts = append(hosts, hostname)
 			continue
 		}
-		if strings.Contains(hostname, "*") {
-			if errs := validation.IsWildcardDNS1123Subdomain(hostname); errs != nil {
-				errors = append(errors, fmt.Errorf("invalid hostname %q: %v", hostname, errs))
+
+		// Validate the hostname.
+		if err := validHostName(hostname); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		if listenerHostname != nil {
+			const wildcardString string = "[a-z0-9]([-a-z0-9]*[a-z0-9])*."
+			lhn := string(*listenerHostname)
+
+			// A "*" hostname matches anything.
+			if lhn == "*" {
+				hosts = append(hosts, hostname)
 				continue
 			}
-		} else {
-			if errs := validation.IsDNS1123Subdomain(hostname); errs != nil {
-				errors = append(errors, fmt.Errorf("invalid listener hostname %q: %v", hostname, errs))
+
+			// If the listener.hostname matches then no need to
+			// do any other validation.
+			if lhn == hostname {
+				hosts = append(hosts, hostname)
 				continue
 			}
+
+			if strings.Contains(lhn, "*") {
```

Comment:
nit: I would switch this to an `else if` joined with the above `if` since they all go together, for clarity, or could do a switch.

Reply:
They all have `continues` in them which showed they exit, but can update.